tools/azsdk-cli/docs/specs/99-repo-tooling-contract.md

```diff
@@ -0,0 +1,322 @@
+
+> Pros: Best of both worlds: a small, stable contract (parameters, exit codes, JSON output) that CLI depends on while letting repo implement details. Repo can evolve implementation without breaking consumers so long as contract remains stable. CLI can provide shared normalization, validation, telemetry, and orchestration, while repo scripts handle repo-local tasks (file paths, tool variants). Easier to test and mock (CLI tests against a known contract; repo tests ensure script adheres to contract).
+
+> Cons: Requires upfront design of the contract and discipline to keep it stable. Still requires versioning decisions (how to roll out contract changes across repos). Some duplication of wrapper code may persist (but minimized if contract is well-designed).
+
+The design proposal below removes some aspects of the contract initially proposed, only keeping the script path and lookup tags.
+
+### Overview
+
+When a command/tool is invoked in `azsdk` that is known to be repository-specific (i.e. the command requires a package path or a call to get the repository root), it should first check a centralized config in that repository. If the config contains an override entry for the command, call the override specified. If not, fall back to the implementation in the CLI/MCP server (or fail/no-op if not implemented).
+
+### Detailed Design
+
+In `eng/azsdk-cli-command-overrides.json` in a language repository, a set of mappings can be defined that
+correspond to CLI commands and/or MCP tools.
+
+```
+[
+  {
+    "tags": ["FooBarCheck", "azsdk_foo_bar", "foobar"],
+    "command": "eng/scripts/FooBar.ps1"
+  }
+]
+```
+
+Any script parameters will be [splatted](https://learn.microsoft.com/powershell/module/microsoft.powershell.core/about/about_splatting) in the script call, and the mapping enabled via the `[Alias]` attribute in the powershell cmdlet. No contract is defined for parameters.
+
+By supporting `[Alias(...)]` script authors can easily coordinate changes to parameter names in `azsdk` without having to
+synchronize releases of the binary with script changes in the local repository.
+Aliases used by older `azsdk` versions can be kept around for backwards compatibility.
+
+To onboard an existing script to be used by the `azsdk` tool, all that needs to be done is add aliases for parameters.
+Alternatively, normalize the keys used by an `azsdk` invocation of the script to match its existing parameter names.
+
+```pwsh
+# eng/scripts/FooBar.ps1
+param(
+  [Alias('fooBarOption', 'foo-bar')]
+  [string] $FooBar
+)
+
+echo $FooBar
+```
+
+On the `azsdk` side, a service will be written to enable the following call patterns:
```

Comment:
> a service will be written to enable the following call patterns:

Sorry I might be lack of context on engsys. By the service, do you mean something like 'tools/azsdk-cli/Azure.Sdk.Tools.Cli/Services/Languages/DotNetLanguageSpecificChecks.cs'? So for update change log tool, Java will write a script in sdk repo to implement the logics and a service called `JavaUpdateChangeLog.cs` to invoke the script.